robotoff/prediction/ocr/grammar.py

```diff
@@ -0,0 +1,155 @@
+    :param lowercase: if True, lowercase names/synonyms
+    :param strip_accent: if True, remove accents from names/synonyms
+    """
+    output = []
+    for node in taxonomy.iter_nodes():
+        if lang not in node.names or node.id in ignore_ids:
+            continue
+
+        names = []
+        if node.names[lang]:
+            # Sometimes it's null
+            names.append(node.names[lang])
+
+        if add_xx and "xx" in node.names:
+            names.append(node.names["xx"])
+
+        if add_synonyms:
+            names += node.synonyms.get(lang, [])
+            if add_xx:
+                names += node.synonyms.get("xx", [])
+
+        names = [normalize_string(name, lowercase, strip_accent) for name in names]
+        # Remove empty items
+        names = [name for name in names if name.strip()]
+        output += [
+            (node.id, name) for name in sorted(set(names), key=lambda x: names.index(x))
+        ]
+
+    return output
+
+
+def generate_terminal_symbols_file(
+    output_path: Path,
+    lang: str,
+    terminal_name: str,
+    taxonomy_type: TaxonomyType,
+    terminal_priority: Optional[int] = None,
+    lowercase: bool = True,
+    strip_accent: bool = True,
+    add_synonyms: bool = True,
+    add_xx: bool = True,
+    ignore_ids: Optional[set[str]] = None,
+    raises: bool = False,
+):
+    """Generate lark terminal symbol file from a taxonomy."""
+    content, name_to_id_mapping = generate_terminal_symbols_text(
+        lang=lang,
+        terminal_name=terminal_name,
+        taxonomy_type=taxonomy_type,
+        terminal_priority=terminal_priority,
+        lowercase=lowercase,
+        strip_accent=strip_accent,
+        add_synonyms=add_synonyms,
+        add_xx=add_xx,
+        ignore_ids=ignore_ids,
+        raises=raises,
+    )
+
+    with output_path.open("w") as f:
+        f.write("// This file has been generated automatically, DO NOT EDIT!\n")
+        f.write(content)
+
+    map_path = output_path.parent / f"{output_path.stem}_map.json"
+    dump_json(map_path, name_to_id_mapping)
+
+
+def generate_terminal_symbols_text(
+    lang: str,
+    terminal_name: str,
+    taxonomy_type: TaxonomyType,
+    terminal_priority: Optional[int] = None,
+    lowercase: bool = True,
+    strip_accent: bool = True,
+    add_synonyms: bool = True,
+    add_xx: bool = True,
+    ignore_ids: Optional[set[str]] = None,
+    raises: bool = True,
+) -> tuple[str, dict[str, list[str]]]:
+    ignore_ids = ignore_ids or set()
+    texts = []
+    taxonomy = get_taxonomy(taxonomy_type.name)
+    seen_set: dict[str, str] = {}
+    previous_node_id = None
+
+    node_id_names = extract_taxonomy_names(
+        lang, taxonomy, ignore_ids, add_synonyms, add_xx, lowercase, strip_accent
+    )
+    for node_id, name in node_id_names:
+        if node_id != previous_node_id:
+            if name in seen_set:
+                error_message = f"item {name} (node {node_id}) already seen for node {seen_set[name]}!"
```

Comment:
I don't really understand this, could you add a comment to explain ? (I don't understand why we only check this if `node_id != previous_node_id`)

Reply:
Me neither :sweat_smile: 
This condition shouldn't exist, we should check if there is a match against any name seen so far